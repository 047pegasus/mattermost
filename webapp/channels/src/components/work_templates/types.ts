// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum ModalState {
    ChannelOnly = 'channel-only',
    Menu = 'menu',
    Customize = 'customize',
    Preview = 'preview',
}
