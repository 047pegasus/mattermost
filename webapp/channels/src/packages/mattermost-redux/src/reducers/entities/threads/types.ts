// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {UsersState} from '@mattermost/types/users';
import type {UserThread, ThreadsState} from '@mattermost/types/threads';

export type ExtraData = {
    threadsToDelete?: UserThread[];
    threads: ThreadsState['threads'];
    currentUserId?: UsersState['currentUserId'];
}
