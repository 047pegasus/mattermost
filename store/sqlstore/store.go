// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package sqlstore

import (
	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"

	// Import enterprise sqlstore backends
	_ "github.com/mattermost/mattermost-server/v5/store/sqlstore/imports"
)
