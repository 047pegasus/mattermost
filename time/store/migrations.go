// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package store

import (
	"github.com/blang/semver"
	"github.com/jmoiron/sqlx"
	"github.com/pkg/errors"

	"github.com/mattermost/mattermost-server/v6/model"
)

type Migration struct {
	fromVersion   semver.Version
	toVersion     semver.Version
	migrationFunc func(sqlx.Ext, *SQLStore) error
}

const MySQLCharset = "DEFAULT CHARACTER SET utf8mb4"

var migrations = []Migration{
	{
		fromVersion: semver.MustParse("0.0.0"),
		toVersion:   semver.MustParse("0.1.0"),
		migrationFunc: func(e sqlx.Ext, sqlStore *SQLStore) error {
			if e.DriverName() == model.DatabaseDriverMysql {
				if _, err := e.Exec(`
					CREATE TABLE IF NOT EXISTS TIME_System (
						SKey VARCHAR(64) PRIMARY KEY,
						SValue VARCHAR(1024) NULL
					)
				` + MySQLCharset); err != nil {
					return errors.Wrapf(err, "failed creating table TIME_System")
				}

				if _, err := e.Exec(`
					CREATE TABLE IF NOT EXISTS TIME_Task (
						ID VARCHAR(26) PRIMARY KEY,
						Title VARCHAR(1024) NOT NULL,
						Complete BOOLEAN NOT NULL,
						BlockID VARCHAR(26) NOT NULL,
						StartAt BIGINT NOT NULL,
						Tags TEXT NOT NULL,
						INDEX TIME_Task_BlockID (BlockID),
					)
				` + MySQLCharset); err != nil {
					return errors.Wrapf(err, "failed creating table TIME_Task")
				}

			} else {

				if _, err := e.Exec(`
					CREATE TABLE IF NOT EXISTS TIME_System (
						SKey VARCHAR(64) PRIMARY KEY,
						SValue VARCHAR(1024) NULL
					);
				`); err != nil {
					return errors.Wrapf(err, "failed creating table TIME_System")
				}

				if _, err := e.Exec(`
					CREATE TABLE IF NOT EXISTS TIME_Task (
						ID TEXT PRIMARY KEY,
						Title TEXT NOT NULL,
						Complete BOOLEAN NOT NULL,
						BlockID TEXT NOT NULL,
						StartAt BIGINT NOT NULL,
						Tags JSON NOT NULL
					);
				`); err != nil {
					return errors.Wrapf(err, "failed creating table TIME_Task")
				}

				if _, err := e.Exec(createPGIndex("TIME_Task_BlockID", "TIME_Task", "BlockID")); err != nil {
					return errors.Wrapf(err, "failed creating index TIME_Task_BlockID ")
				}
			}

			return nil
		},
	},
}
