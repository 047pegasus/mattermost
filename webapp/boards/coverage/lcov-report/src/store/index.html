
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src/store</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> src/store</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">38.61% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>446/1155</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">18.46% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>70/379</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">34.16% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>150/439</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">39.61% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>368/929</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="attachments.ts"><a href="attachments.ts.html">attachments.ts</a></td>
	<td data-value="23.63" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 23%"></div><div class="cover-empty" style="width: 77%"></div></div>
	</td>
	<td data-value="23.63" class="pct low">23.63%</td>
	<td data-value="55" class="abs low">13/55</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="15" class="abs low">0/15</td>
	<td data-value="36.84" class="pct low">36.84%</td>
	<td data-value="19" class="abs low">7/19</td>
	<td data-value="23.91" class="pct low">23.91%</td>
	<td data-value="46" class="abs low">11/46</td>
	</tr>

<tr>
	<td class="file low" data-value="boards.ts"><a href="boards.ts.html">boards.ts</a></td>
	<td data-value="32.53" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 32%"></div><div class="cover-empty" style="width: 68%"></div></div>
	</td>
	<td data-value="32.53" class="pct low">32.53%</td>
	<td data-value="166" class="abs low">54/166</td>
	<td data-value="22.38" class="pct low">22.38%</td>
	<td data-value="67" class="abs low">15/67</td>
	<td data-value="35.38" class="pct low">35.38%</td>
	<td data-value="65" class="abs low">23/65</td>
	<td data-value="30.93" class="pct low">30.93%</td>
	<td data-value="139" class="abs low">43/139</td>
	</tr>

<tr>
	<td class="file low" data-value="cards.ts"><a href="cards.ts.html">cards.ts</a></td>
	<td data-value="36.73" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 36%"></div><div class="cover-empty" style="width: 64%"></div></div>
	</td>
	<td data-value="36.73" class="pct low">36.73%</td>
	<td data-value="245" class="abs low">90/245</td>
	<td data-value="19.2" class="pct low">19.2%</td>
	<td data-value="151" class="abs low">29/151</td>
	<td data-value="29.48" class="pct low">29.48%</td>
	<td data-value="78" class="abs low">23/78</td>
	<td data-value="38.42" class="pct low">38.42%</td>
	<td data-value="203" class="abs low">78/203</td>
	</tr>

<tr>
	<td class="file medium" data-value="channels.ts"><a href="channels.ts.html">channels.ts</a></td>
	<td data-value="59.09" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 59%"></div><div class="cover-empty" style="width: 41%"></div></div>
	</td>
	<td data-value="59.09" class="pct medium">59.09%</td>
	<td data-value="22" class="abs medium">13/22</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="9" class="abs low">3/9</td>
	<td data-value="71.42" class="pct medium">71.42%</td>
	<td data-value="14" class="abs medium">10/14</td>
	</tr>

<tr>
	<td class="file medium" data-value="clientConfig.ts"><a href="clientConfig.ts.html">clientConfig.ts</a></td>
	<td data-value="68.75" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 68%"></div><div class="cover-empty" style="width: 32%"></div></div>
	</td>
	<td data-value="68.75" class="pct medium">68.75%</td>
	<td data-value="16" class="abs medium">11/16</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="44.44" class="pct low">44.44%</td>
	<td data-value="9" class="abs low">4/9</td>
	<td data-value="76.92" class="pct medium">76.92%</td>
	<td data-value="13" class="abs medium">10/13</td>
	</tr>

<tr>
	<td class="file low" data-value="comments.ts"><a href="comments.ts.html">comments.ts</a></td>
	<td data-value="34.84" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 34%"></div><div class="cover-empty" style="width: 66%"></div></div>
	</td>
	<td data-value="34.84" class="pct low">34.84%</td>
	<td data-value="66" class="abs low">23/66</td>
	<td data-value="37.5" class="pct low">37.5%</td>
	<td data-value="24" class="abs low">9/24</td>
	<td data-value="65" class="pct medium">65%</td>
	<td data-value="20" class="abs medium">13/20</td>
	<td data-value="35.08" class="pct low">35.08%</td>
	<td data-value="57" class="abs low">20/57</td>
	</tr>

<tr>
	<td class="file low" data-value="contents.ts"><a href="contents.ts.html">contents.ts</a></td>
	<td data-value="37.2" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 37%"></div><div class="cover-empty" style="width: 63%"></div></div>
	</td>
	<td data-value="37.2" class="pct low">37.2%</td>
	<td data-value="86" class="abs low">32/86</td>
	<td data-value="30.55" class="pct low">30.55%</td>
	<td data-value="36" class="abs low">11/36</td>
	<td data-value="59.09" class="pct medium">59.09%</td>
	<td data-value="22" class="abs medium">13/22</td>
	<td data-value="38.66" class="pct low">38.66%</td>
	<td data-value="75" class="abs low">29/75</td>
	</tr>

<tr>
	<td class="file medium" data-value="globalError.ts"><a href="globalError.ts.html">globalError.ts</a></td>
	<td data-value="60" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 60%"></div><div class="cover-empty" style="width: 40%"></div></div>
	</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="15" class="abs medium">9/15</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="72.72" class="pct medium">72.72%</td>
	<td data-value="11" class="abs medium">8/11</td>
	</tr>

<tr>
	<td class="file medium" data-value="globalTemplates.ts"><a href="globalTemplates.ts.html">globalTemplates.ts</a></td>
	<td data-value="53.33" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 53%"></div><div class="cover-empty" style="width: 47%"></div></div>
	</td>
	<td data-value="53.33" class="pct medium">53.33%</td>
	<td data-value="15" class="abs medium">8/15</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="8" class="abs low">2/8</td>
	<td data-value="58.33" class="pct medium">58.33%</td>
	<td data-value="12" class="abs medium">7/12</td>
	</tr>

<tr>
	<td class="file high" data-value="hooks.ts"><a href="hooks.ts.html">hooks.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="20" class="abs high">20/20</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="19" class="abs high">19/19</td>
	</tr>

<tr>
	<td class="file low" data-value="initialLoad.ts"><a href="initialLoad.ts.html">initialLoad.ts</a></td>
	<td data-value="42.85" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 42%"></div><div class="cover-empty" style="width: 58%"></div></div>
	</td>
	<td data-value="42.85" class="pct low">42.85%</td>
	<td data-value="35" class="abs low">15/35</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="35.71" class="pct low">35.71%</td>
	<td data-value="14" class="abs low">5/14</td>
	<td data-value="37.03" class="pct low">37.03%</td>
	<td data-value="27" class="abs low">10/27</td>
	</tr>

<tr>
	<td class="file medium" data-value="language.ts"><a href="language.ts.html">language.ts</a></td>
	<td data-value="57.89" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 57%"></div><div class="cover-empty" style="width: 43%"></div></div>
	</td>
	<td data-value="57.89" class="pct medium">57.89%</td>
	<td data-value="19" class="abs medium">11/19</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="36.36" class="pct low">36.36%</td>
	<td data-value="11" class="abs low">4/11</td>
	<td data-value="62.5" class="pct medium">62.5%</td>
	<td data-value="16" class="abs medium">10/16</td>
	</tr>

<tr>
	<td class="file medium" data-value="limits.ts"><a href="limits.ts.html">limits.ts</a></td>
	<td data-value="57.89" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 57%"></div><div class="cover-empty" style="width: 43%"></div></div>
	</td>
	<td data-value="57.89" class="pct medium">57.89%</td>
	<td data-value="19" class="abs medium">11/19</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="20" class="pct low">20%</td>
	<td data-value="10" class="abs low">2/10</td>
	<td data-value="76.92" class="pct medium">76.92%</td>
	<td data-value="13" class="abs medium">10/13</td>
	</tr>

<tr>
	<td class="file medium" data-value="searchText.ts"><a href="searchText.ts.html">searchText.ts</a></td>
	<td data-value="77.77" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 77%"></div><div class="cover-empty" style="width: 23%"></div></div>
	</td>
	<td data-value="77.77" class="pct medium">77.77%</td>
	<td data-value="9" class="abs medium">7/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="60" class="pct medium">60%</td>
	<td data-value="5" class="abs medium">3/5</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	</tr>

<tr>
	<td class="file low" data-value="sidebar.ts"><a href="sidebar.ts.html">sidebar.ts</a></td>
	<td data-value="13.04" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 13%"></div><div class="cover-empty" style="width: 87%"></div></div>
	</td>
	<td data-value="13.04" class="pct low">13.04%</td>
	<td data-value="92" class="abs low">12/92</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="21" class="abs low">0/21</td>
	<td data-value="5.4" class="pct low">5.4%</td>
	<td data-value="37" class="abs low">2/37</td>
	<td data-value="15.27" class="pct low">15.27%</td>
	<td data-value="72" class="abs low">11/72</td>
	</tr>

<tr>
	<td class="file low" data-value="teams.ts"><a href="teams.ts.html">teams.ts</a></td>
	<td data-value="47.05" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 47%"></div><div class="cover-empty" style="width: 53%"></div></div>
	</td>
	<td data-value="47.05" class="pct low">47.05%</td>
	<td data-value="51" class="abs low">24/51</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="24" class="abs low">8/24</td>
	<td data-value="48.57" class="pct low">48.57%</td>
	<td data-value="35" class="abs low">17/35</td>
	</tr>

<tr>
	<td class="file low" data-value="users.ts"><a href="users.ts.html">users.ts</a></td>
	<td data-value="45.68" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 45%"></div><div class="cover-empty" style="width: 55%"></div></div>
	</td>
	<td data-value="45.68" class="pct low">45.68%</td>
	<td data-value="116" class="abs low">53/116</td>
	<td data-value="27.27" class="pct low">27.27%</td>
	<td data-value="22" class="abs low">6/22</td>
	<td data-value="31.57" class="pct low">31.57%</td>
	<td data-value="57" class="abs low">18/57</td>
	<td data-value="47.67" class="pct low">47.67%</td>
	<td data-value="86" class="abs low">41/86</td>
	</tr>

<tr>
	<td class="file low" data-value="views.ts"><a href="views.ts.html">views.ts</a></td>
	<td data-value="33.33" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 33%"></div><div class="cover-empty" style="width: 67%"></div></div>
	</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="102" class="abs low">34/102</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="23" class="abs low">0/23</td>
	<td data-value="35.89" class="pct low">35.89%</td>
	<td data-value="39" class="abs low">14/39</td>
	<td data-value="30.86" class="pct low">30.86%</td>
	<td data-value="81" class="abs low">25/81</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at Fri Mar 10 2023 07:37:49 GMT-0600 (Central Standard Time)
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    