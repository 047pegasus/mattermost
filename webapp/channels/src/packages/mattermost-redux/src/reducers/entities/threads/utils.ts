// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {GenericAction} from 'mattermost-redux/types/actions';
import type {IDMappedObjects, RelationOneToMany} from '@mattermost/types/utilities';
import type {UserThread} from '@mattermost/types/threads';

import type {ExtraData} from './types';

type State = RelationOneToMany<{id: string}, UserThread>;

// return true only if it's 'newer' than other threads
// older threads will be added by scrolling so no need to manually add.
// furthermore manually adding older thread will BREAK pagination
export function shouldAddThreadId(ids: Array<UserThread['id']>, thread: UserThread, threads: IDMappedObjects<UserThread>) {
    return ids.some((id) => {
        const t = threads![id];
        return thread.last_reply_at > t.last_reply_at;
    });
}

// adds thread to single team
export function handleReceivedThread<S extends State>(state: S, thread: UserThread, key: string, extra: ExtraData): S {
    const nextSet = new Set(state[key] || []);

    // thread exists in state
    if (nextSet.has(thread.id)) {
        return state;
    }

    // check if thread is newer than any of the existing threads
    const shouldAdd = shouldAddThreadId([...nextSet], thread, extra.threads);

    if (shouldAdd) {
        nextSet.add(thread.id);

        return {
            ...state,
            [key]: [...nextSet],
        };
    }

    return state;
}

export function handlePostRemoved<S extends State>(state: S, action: GenericAction): S {
    const post = action.data;
    if (post.root_id) {
        return state;
    }

    const keys = Object.keys(state).
        filter((id) => state[id].indexOf(post.id) !== -1);

    if (!keys?.length) {
        return state;
    }

    const newState: State = {};

    for (let i = 0; i < keys.length; i++) {
        const key = keys[i];
        const index = state[key].indexOf(post.id);

        newState[key] = [
            ...state[key].slice(0, index),
            ...state[key].slice(index + 1),
        ];
    }

    return {
        ...state,
        ...newState,
    };
}

export function handleReceiveThreads<S extends State>(state: S, action: GenericAction, key: string): S {
    const nextSet = new Set(state[key] || []);

    if (action.data.threads.length === 0) {
        return state;
    }

    action.data.threads.forEach((thread: UserThread) => {
        nextSet.add(thread.id);
    });

    return {
        ...state,
        [key]: [...nextSet],
    };
}

// add the thread only if it's 'newer' than other threads
// older threads will be added by scrolling so no need to manually add.
// furthermore manually adding older thread will BREAK pagination
export function handleFollowChanged<S extends State>(state: S, action: GenericAction, key: string, extra: ExtraData): S {
    const {id, following} = action.data;
    const nextSet = new Set(state[key] || []);

    const thread = extra.threads[id];

    if (!thread) {
        return state;
    }

    // thread exists in state
    if (nextSet.has(id)) {
        // remove it if we unfollowed
        if (!following) {
            nextSet.delete(id);
            return {
                ...state,
                [key]: [...nextSet],
            };
        }
        return state;
    }

    // check if thread is newer than any of the existing threads
    const shouldAdd = shouldAddThreadId([...nextSet], thread, extra.threads);

    if (shouldAdd && following) {
        nextSet.add(thread.id);

        return {
            ...state,
            [key]: [...nextSet],
        };
    }

    return state;
}
