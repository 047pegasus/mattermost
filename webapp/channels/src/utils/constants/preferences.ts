// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export const CATEGORY_CHANNEL_OPEN_TIME = 'channel_open_time';
export const CATEGORY_DIRECT_CHANNEL_SHOW = 'direct_channel_show';
export const CATEGORY_GROUP_CHANNEL_SHOW = 'group_channel_show';
export const CATEGORY_DISPLAY_SETTINGS = 'display_settings';
export const CATEGORY_SIDEBAR_SETTINGS = 'sidebar_settings';
export const CATEGORY_ADVANCED_SETTINGS = 'advanced_settings';
export const TUTORIAL_STEP = 'tutorial_step';
export const TUTORIAL_STEP_AUTO_TOUR_STATUS = 'tutorial_step_auto_tour_status';
export const CRT_TUTORIAL_TRIGGERED = 'crt_tutorial_triggered';
export const CRT_TUTORIAL_AUTO_TOUR_STATUS = 'crt_tutorial_auto_tour_status';
export const CRT_TUTORIAL_STEP = 'crt_tutorial_step';
export const EXPLORE_OTHER_TOOLS_TUTORIAL_STEP = 'explore_other_tools_step';
export const CRT_THREAD_PANE_STEP = 'crt_thread_pane_step';
export const CHANNEL_DISPLAY_MODE = 'channel_display_mode';
export const CHANNEL_DISPLAY_MODE_CENTERED = 'centered';
export const CHANNEL_DISPLAY_MODE_FULL_SCREEN = 'full';
export const CHANNEL_DISPLAY_MODE_DEFAULT = 'full';
export const MESSAGE_DISPLAY = 'message_display';
export const MESSAGE_DISPLAY_CLEAN = 'clean';
export const MESSAGE_DISPLAY_COMPACT = 'compact';
export const MESSAGE_DISPLAY_DEFAULT = 'clean';
export const COLORIZE_USERNAMES = 'colorize_usernames';
export const COLORIZE_USERNAMES_DEFAULT = 'true';
export const COLLAPSED_REPLY_THREADS = 'collapsed_reply_threads';
export const COLLAPSED_REPLY_THREADS_OFF = 'off';
export const COLLAPSED_REPLY_THREADS_ON = 'on';
export const CLICK_TO_REPLY = 'click_to_reply';
export const CLICK_TO_REPLY_DEFAULT = 'true';
export const COLLAPSED_REPLY_THREADS_FALLBACK_DEFAULT = 'off';
export const LINK_PREVIEW_DISPLAY = 'link_previews';
export const LINK_PREVIEW_DISPLAY_DEFAULT = 'true';
export const COLLAPSE_DISPLAY = 'collapse_previews';
export const COLLAPSE_DISPLAY_DEFAULT = 'false';
export const AVAILABILITY_STATUS_ON_POSTS = 'availability_status_on_posts';
export const AVAILABILITY_STATUS_ON_POSTS_DEFAULT = 'true';
export const USE_MILITARY_TIME = 'use_military_time';
export const USE_MILITARY_TIME_DEFAULT = 'false';
export const UNREAD_SCROLL_POSITION = 'unread_scroll_position';
export const UNREAD_SCROLL_POSITION_START_FROM_LEFT = 'start_from_left_off';
export const UNREAD_SCROLL_POSITION_START_FROM_NEWEST = 'start_from_newest';
export const CATEGORY_THEME = 'theme';
export const CATEGORY_FLAGGED_POST = 'flagged_post';
export const CATEGORY_NOTIFICATIONS = 'notifications';
export const EMAIL_INTERVAL = 'email_interval';
export const INTERVAL_IMMEDIATE = 30; // "immediate" is a 30 second interval
export const INTERVAL_FIFTEEN_MINUTES = 15 * 60;
export const INTERVAL_HOUR = 60 * 60;
export const INTERVAL_NEVER = 0;
export const NAME_NAME_FORMAT = 'name_format';
export const CATEGORY_SYSTEM_NOTICE = 'system_notice';
export const RECOMMENDED_NEXT_STEPS = 'recommended_next_steps';
export const TEAMS_ORDER = 'teams_order';
export const CLOUD_UPGRADE_BANNER = 'cloud_upgrade_banner';
export const CLOUD_TRIAL_BANNER = 'cloud_trial_banner';
export const START_TRIAL_MODAL = 'start_trial_modal';
export const ADMIN_CLOUD_UPGRADE_PANEL = 'admin_cloud_upgrade_panel';
export const CATEGORY_EMOJI = 'emoji';
export const EMOJI_SKINTONE = 'emoji_skintone';
export const ONE_CLICK_REACTIONS_ENABLED = 'one_click_reactions_enabled';
export const ONE_CLICK_REACTIONS_ENABLED_DEFAULT = 'true';
export const CLOUD_TRIAL_END_BANNER = 'cloud_trial_end_banner';
export const CLOUD_USER_EPHEMERAL_INFO = 'cloud_user_ephemeral_info';
export const CATEGORY_CLOUD_LIMITS = 'cloud_limits';
export const THREE_DAYS_LEFT_TRIAL_MODAL = 'three_days_left_trial_modal';

// For one off things that have a special, attention-grabbing UI until you interact with them
export const TOUCHED = 'touched';

// Category for actions/interactions that will happen just once
export const UNIQUE = 'unique';

// A/B test preference value
export const AB_TEST_PREFERENCE_VALUE = 'ab_test_preference_value';

export const RECENT_EMOJIS = 'recent_emojis';
export const ONBOARDING = 'onboarding';
export const ADVANCED_TEXT_EDITOR = 'advanced_text_editor';

export const FORWARD_POST_VIEWED = 'forward_post_viewed';
export const HIDE_POST_FILE_UPGRADE_WARNING = 'hide_post_file_upgrade_warning';
export const SHOWN_LIMITS_REACHED_ON_LOGIN = 'shown_limits_reached_on_login';
export const USE_CASE = 'use_case';
export const DELINQUENCY_MODAL_CONFIRMED = 'delinquency_modal_confirmed';
export const CONFIGURATION_BANNERS = 'configuration_banners';
export const NOTIFY_ADMIN_REVOKE_DOWNGRADED_WORKSPACE = 'admin_revoke_downgraded_instance';
export const OVERAGE_USERS_BANNER = 'overage_users_banner';
export const TO_CLOUD_YEARLY_PLAN_NUDGE = 'to_cloud_yearly_plan_nudge';
export const TO_PAID_PLAN_NUDGE = 'to_paid_plan_nudge';
