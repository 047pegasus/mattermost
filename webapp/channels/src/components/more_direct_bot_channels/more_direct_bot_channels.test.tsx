// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {GlobalState} from '@mattermost/types/store';

import {ModalIdentifiers} from 'utils/constants';

import MoreDirectBotChannelsModal from './';

const mockDispatch = jest.fn();
let mockState: GlobalState;

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux') as typeof import('react-redux'),
    useSelector: (selector: (state: typeof mockState) => unknown) => selector(mockState),
    useDispatch: () => mockDispatch,
}));

describe('components/more_direct_bot_channels', () => {
    mockState = {
        views: {
            modals: {
                modalState: {
                    [ModalIdentifiers.MORE_DIRECT_BOT_CHANNELS]: {
                        open: true,
                    },
                },
            },
        },
    } as unknown as GlobalState;

    test('should match snapshot', () => {
        expect(
            shallow(
                <MoreDirectBotChannelsModal/>,
            ),
        ).toMatchSnapshot();
    });
});
