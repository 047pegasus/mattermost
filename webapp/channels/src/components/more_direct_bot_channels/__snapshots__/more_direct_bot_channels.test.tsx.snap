// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/more_direct_bot_channels should match snapshot 1`] = `
<LegacyGenericModal
  ariaLabel="Apps"
  autoCloseOnCancelButton={true}
  autoCloseOnConfirmButton={true}
  bodyPadding={false}
  className="MoreDirectBotChannels-modal"
  compassDesign={true}
  enforceFocus={true}
  id="MoreDirectBotChannels-modal"
  keyboardEscape={true}
  modalHeaderText="Apps"
  onExited={[Function]}
  show={true}
>
  <div
    style={
      Object {
        "height": 200,
      }
    }
  />
</LegacyGenericModal>
`;
